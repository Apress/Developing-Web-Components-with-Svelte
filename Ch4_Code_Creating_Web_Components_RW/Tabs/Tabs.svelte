<svelte:options tag="cobalt-tabs" />

<script>
  import tabItems from "./tabsdata.json";
  export let activeTabValue = "0";
  export let items;
  export let vertical = false;

  if (items == null) {
    items = tabItems;
  }

  const handleClick = (tabValue) => () => (activeTabValue = tabValue);
</script>

<div class="cobalt" class:vertical>
  <ul>
    {#each Object.entries(items) as [id, name]}
      <li class={activeTabValue === id ? "active" : ""}>
        <span on:click={handleClick(id)}>
          {JSON.stringify(items[id].name).replace(/['"]+/g, "")}
        </span>
      </li>
    {/each}
  </ul>

  <div class="content">
    {#each Object.entries(items) as [id, text]}
      {#if activeTabValue === id}
        {JSON.stringify(items[id].text).replace(/['"]+/g, "")}
      {/if}
    {/each}
  </div>
</div>

<style>
  .cobalt {
    display: block;
  }

  ul {
    display: flex;
    flex-wrap: wrap;
    padding-left: 0;
    margin-bottom: 0;
    list-style: none;
    border-bottom: 1px solid #dee2e6;
  }

  .content {
    padding: 10px;
  }

  span {
    border: 1px solid transparent;
    border-top-left-radius: 4px;
    border-top-right-radius: 4px;
    display: block;
    padding: 8px 16px;
    cursor: pointer;
  }

  span:hover {
    border-color: #e9ecef #e9ecef #dee2e6;
    background-color: #9393ff;
  }

  li.active > span {
    color: #ffffff;
    background-color: #19247c;
    border-color: #dee2e6 #dee2e6 #fff;
  }

  /* variant */
  div.cobalt.vertical {
    display: flex;
  }

  div.cobalt.vertical ul {
    flex-direction: column;
    border-bottom: none;
    align-self: baseline;
    margin-top: 0;
  }

  div.cobalt.vertical div.content {
    width: 300px;
    height: 200px;
  }

  div.cobalt.vertical ul li span {
    border-top-left-radius: 0;
    border-top-right-radius: 0;
  }
</style>
