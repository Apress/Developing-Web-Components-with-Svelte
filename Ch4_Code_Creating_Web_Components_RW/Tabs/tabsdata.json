[
  {"id": 1, "name": "Name 1", "text": "Text 1"},
  {"id": 2, "name": "Name 2", "text": "Text 2"},
  {"id": 3, "name": "Name 3", "text": "Text 3"}
]
