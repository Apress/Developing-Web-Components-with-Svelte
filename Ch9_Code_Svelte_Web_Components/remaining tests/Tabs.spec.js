import { fireEvent, render, screen } from '@testing-library/svelte';
import Tabs from "../src/lib/Tabs/Tabs.svelte";

describe("Tests for Tabs", () => {
  const props = {
    vertical: false,
    items: [
      { "id": 1, "name": "Tab 1", "text": "This is a test"},
      { "id": 2, "name": "Tab 2", "text": "Here is tab 2"},
      { "id": 3, "name": "Tab 3", "text": "And this is tab 3"},
    ],
  };

  it("should render properly", () => {
    const result = render(Tabs, props );
    expect(() => result).not.toThrow();
  });

  it("should change text if clicking on tab 2", async()=> {
    const result = render(Tabs, props );
    const tabResult = await(screen.getAllByRole("")[0]);
    const tabHeader = result.getByTextId("tabHeader")[1];
    await fireEvent.click(tabHeader);
    expect(screen.getByTextId("tabContent")[1]).toBe("Here is tab 2")
  })

  it("get a snapshot of component", () => {
    const tree = render(Tabs, props);
    expect(tree).toMatchSnapshot();
  })
});