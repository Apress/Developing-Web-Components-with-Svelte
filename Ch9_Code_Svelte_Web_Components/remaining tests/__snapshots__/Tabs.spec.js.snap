// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for Tabs get a snapshot of component 1`] = `
{
  "component": Tabs {
    "$$": {
      "after_update": [],
      "before_update": [],
      "bound": {},
      "callbacks": {},
      "context": Map {},
      "ctx": [
        "0",
        [
          {
            "id": 1,
            "name": "Tab 1",
            "text": "This is a test",
          },
          {
            "id": 2,
            "name": "Tab 2",
            "text": "Here is tab 2",
          },
          {
            "id": 3,
            "name": "Tab 3",
            "text": "And this is tab 3",
          },
        ],
        false,
        [Function],
      ],
      "dirty": [
        -1,
      ],
      "fragment": {
        "c": [Function],
        "d": [Function],
        "i": [Function],
        "l": [Function],
        "m": [Function],
        "o": [Function],
        "p": [Function],
      },
      "not_equal": [Function],
      "on_destroy": [
        [Function],
      ],
      "on_disconnect": [],
      "on_mount": [],
      "props": {
        "activeTabValue": 0,
        "items": 1,
        "vertical": 2,
      },
      "root": <div>
        <div
          class="cobalt svelte-11nojaq"
        >
          <ul
            class="svelte-11nojaq"
          >
            <li
              class="active svelte-11nojaq"
              data-testid="tabHeader"
            >
              <span
                class="svelte-11nojaq"
              >
                Tab 1
              </span>
               
            </li>
            <li
              class=" svelte-11nojaq"
              data-testid="tabHeader"
            >
              <span
                class="svelte-11nojaq"
              >
                Tab 2
              </span>
               
            </li>
            <li
              class=" svelte-11nojaq"
              data-testid="tabHeader"
            >
              <span
                class="svelte-11nojaq"
              >
                Tab 3
              </span>
               
            </li>
          </ul>
           
          <div
            class="content svelte-11nojaq"
            data-testid="tabContent"
          >
            This is a test
            
            
            
          </div>
        </div>
      </div>,
      "skip_bound": false,
      "update": [Function],
    },
    "$$set": [Function],
    "$capture_state": [Function],
    "$inject_state": [Function],
  },
  "container": <body>
    <div>
      <div
        class="cobalt svelte-11nojaq"
      >
        <ul
          class="svelte-11nojaq"
        >
          <li
            class="active svelte-11nojaq"
            data-testid="tabHeader"
          >
            <span
              class="svelte-11nojaq"
            >
              Tab 1
            </span>
             
          </li>
          <li
            class=" svelte-11nojaq"
            data-testid="tabHeader"
          >
            <span
              class="svelte-11nojaq"
            >
              Tab 2
            </span>
             
          </li>
          <li
            class=" svelte-11nojaq"
            data-testid="tabHeader"
          >
            <span
              class="svelte-11nojaq"
            >
              Tab 3
            </span>
             
          </li>
        </ul>
         
        <div
          class="content svelte-11nojaq"
          data-testid="tabContent"
        >
          This is a test
          
          
          
        </div>
      </div>
    </div>
  </body>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
